data_structures/linked_list/circular_doubly_linked_list.c

```diff
@@ -0,0 +1,322 @@
+ */
+ListNode *delete_from_tail(ListNode *head)
+{
+    if (head == NULL)
+    {
+        printf("The list is empty\n");
+        return head;
+    }
+
+    ListNode *temp1, *temp2;
+    temp1 = head;
+    temp2 = temp1->prev;
+    if (temp1 == temp2)
+    {
+        free(temp2);
+        head = NULL;
+        return head;
+    }
+    (temp2->prev)->next = temp1;
+    temp1->prev = temp2->prev;
+    free(temp2);
+    return head;
+}
+
+/**
+ * @brief The function to reverse the linked list
+ *
+ * @param head              start pointer of list
+ * @return ListNode*        pointer to the list node after reversing linked list
+ */
+ListNode *reverse_list(ListNode *head)
+{
+    if (!head)
+    {
+        printf("The list is empty\n");
+        return NULL;
+    }
+    ListNode *new_head = NULL;
+    ListNode *last = head->prev;
+    ListNode *curr = last, *prev;
+
+    // traverse list in backward direction
+    while (curr->prev != last)
+    {
+        prev = curr->prev;
+        new_head = insert_at_tail(new_head, curr->value);
+        curr = prev;
+    }
+    new_head = insert_at_tail(new_head, curr->value);
+    return new_head;
+}
+
+/**
+ * @brief The function that will return current size of list
+ *
+ * @param head      start pointer of list
+ * @return int      size of list
+ */
+int getsize(ListNode *head)
+{
+    if (!head)
+    {
+        return 0;
+    }
+    int size = 1;
+    ListNode *temp = head->next;
+    while (temp != head)
+    {
+        temp = temp->next;
+        size++;
+    }
+    return size;
+}
+
+/**
+ * @brief   Display list function
+ * @param   head   start pointer of list
+ * @returns void
+ */
+
+void display_list(ListNode *head)
+{
+    printf("\nContents of your linked list: ");
+    ListNode *temp;
+    temp = head;
+    if (head != NULL)
+    {
+        while (temp->next != head)
+        {
+            printf("%lld -> ", temp->value);
+            temp = temp->next;
+        }
+        if (temp->next == head)
+        {
+            printf("%lld", temp->value);
+        }
+    }
+    else
+    {
+        printf("The list is empty");
+    }
+    printf("\n");
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{
+    int arr[5] = {1, 2, 3, 4, 5};
+    ListNode *arrList = NULL;
+    for (int i = 0; i < 5; i++)
+    {
+        arrList = insert_at_tail(arrList, arr[i]);
+    }
+    ListNode *validationList = NULL;
+    validationList = insert_at_head(validationList, 5);
+    validationList = insert_at_head(validationList, 4);
+    validationList = insert_at_head(validationList, 3);
+    validationList = insert_at_head(validationList, 2);
+    validationList = insert_at_head(validationList, 1);
+
+    ListNode *temp1 = validationList;
+    ListNode *temp2 = arrList;
+    while (temp1->next != validationList)
+    {
+        assert(temp1->value == temp2->value);
+        temp1 = temp1->next;
+        temp2 = temp2->next;
+    }
+
+    // Creating an empty list
+    ListNode *head = NULL;
```

Comment:
Starting from this line all these previous testing code becomes mostly unnecessary, if you want to let the user see what's going on, just display `validationList` or `arrList` should be good enough

Reply:
Actually, I kept it as I thought inserting and deleting separately makes more sense.